rfcs/taskforce/0002-network-utility.md

```diff
@@ -0,0 +1,108 @@
+# Network Utility Tool
+
+## Goal
+
+When spinning up a network there are a few operations that can and have been automated to decrease iteration speed on testing newer versions of the network.
+
+1. Keyset Generation (Account Creation)
+2. Genesis Ledger Creation
+3. Distribution of custom network configuration and keys
+4. Deployment of infrastructure with updated configuration
+5. Monitoring the health of the network and debuging issues / crashes
+
+This document proposes a tool that unifies these processes by collecting tools already written for these purposes and packaging them in a simple CLI and JS API.
+
+# Design
+
+## Existing System
+
+For these processes we mainly use the [testnet-keys.py](https://github.com/CodaProtocol/coda-automation/blob/master/scripts/testnet-keys.py) script in coda-automation.
+
+1. Keypair generation calls `coda advanced-generate keypair` generate new accounts
+2. Genesis ledger creation happens using custom logic inside this script
+3. Distribution of configuration uses Kubernetes secrets
+4. Deployment consists of a variety of terraform configurations targeting Google Kubernetes Engine
+5. We have a variety of charts and metrics for monitoring macro network health but debugging specific issues is often manual
+
+### Improvements
+
+While this system is effective, there are a variety of potential improvements to this process.
+
+- Keypair generation depends on the daemon
+- Genesis ledger logic could be cleaned up and made more flexible to support a wider variety of configuraitons
+- These tools could be packaged together with a JavaScript API for extending functionality
+- The internal data structures for storing these keypairs, genesis configuration and distribution are not well defined and could be more robust
+- Tools for deployment and monitoring are seperate and these could be packaged in one simple workflow that uses the same underlying functionality
+
+## Proposal
+
+All of these features should be included in a simple command line tool that also exposes a TypeScript API. This service will be written in Reason with TypeScript bindings.
+
+### Keypair Management
+
+For new keypair generation, we can use the new client-sdk which is a much lighter dependency and has a clean JavaScript interface.
+
+```
+type keypair = {
+  publicKey: string,
+  privateKey: string
+}
+```
+
+However there are also existing keys which we'll need to be able to reference by public key alone since we don't control the private keys. These include community owned wallets for inclusion in the genesis ledger.The basic type of a keypair consists of the following:
+
+### Keyset Management
```

Comment:
Keysets should be stored externally in some shared secrets manager (or secure flat file system) so that we do not need to ship them between developer machines. Right now, I will generate a keyset and genesis ledger locally on my machine, and commit the ledger to the coda repo so it's included in the build. If I want to allow another developer to deploy that network, I need to zip up my keyset and send it to them. Ideally, keysets would be identified by names and cached externally from our local machines. Then, when someone deploys, the deployment system can pull the keyset from the remote cache instead of generate them locally. Without this, we would need to continue shipping around zip files or attach a new genesis ledger to each deployment we build.

Reply:
As far as an initial implementation is concerned, I think outputting to local disk and AWS Secrets Manager would suffice and stubbing it out so further backends could be built in is useful.